opm/autodiff/FlowMain.hpp

```diff
@@ -252,9 +253,67 @@ namespace Opm
 #endif
         }
 
+        /// checks cartesian adjacency of global indices g1 and g2
+        bool cartesianAdjacent(const Grid& grid, size_t g1, size_t g2) {
+            // we need cartDims from UgGridHelpers
+            using namespace UgGridHelpers;
 
+            auto diff = std::abs(g1 - g2);
 
+            const int * dimens = cartDims(grid);
+            if (diff == 1)
+               return true;
+            if (diff == dimens[0])
+               return true;
+            if (diff == dimens[0] * dimens[1])
+               return true;
 
+            return false;
+        }
+
+        /*!
+         * \brief Get the NNC structure of the given grid.
+         *
+         * Get cell adjacencies beyond Cartesian neighborhoods.  The
+         * transmissivity is given as -1d (unless trans vector is provided)
+         * hence NNC "structure".
+         *
+         * The trans vector is indexed by face number as it is in grid.
+         */
+        NNC nncStructure(const Grid& grid, const std::vector<double>& trans = std::vector<double>()) {
+            // we use numFaces, numCells, cell2Faces, globalCell from UgGridHelpers
+            using namespace UgGridHelpers;
+            NNC nnc;
+
+            typedef std::vector<int> nb; // neighbors
+            size_t num_faces = numFaces(grid);
+            if (trans.size() > 0 && trans.size() != num_faces)
+                throw std::logic_error(
+                            "non-empty trans vector with wrong dimension.");
+
+            std::vector<nb> faces(num_faces); // a "2 by f matrix" face2cells
+            for (int cell = 0; cell < numCells(grid); cell++) { // active cell idx
+                for (size_t face_id : cell2Faces(grid)[cell]) {
+                    faces[face_id].push_back(cell);
+                }
+            }
+
+            for (size_t i = 0; i < num_faces; ++i) {
+                std::vector<int> face = faces[i]; // the two cells on each side
```

Comment:
This is a copy involving heap allocation, should be avoided in an inner loop. Now it does not matter since you will use the grid-provided information instead...